### DISCUSSION TITLE
Reactivity Interoperability

### DISCUSSION THREAD

**Discussion by 3Shain** [vuejs/rfcs#423] [RFC Discussions]
- Start Date: 2021-12-16
- Target Major Version: 3.x
- Reference Issues: (fill in existing related issues, if any)
- Implementation PR: (leave this empty)

# Summary

Introducing a new `addReactivityInterop()` API for `@vue/reactivity`. This feature adds support for (one-way) reactivity interoperability with third party libraries.

# Basic example

```ts
// file: store.js

// use mobx as a demostration of external reactivity source
import { Reaction, makeAutoObservable } from 'mobx'

// register MobX
// this step should usually be performed by _library authors_.
let id = 0
addReactivityInterop((fn, trigger) => {
  const reaction = new Reaction(`externalSource@${++id}`, trigger)
  return {
    track: (x) => {
      let next
      reaction.track(() => (next = fn(x)))
      return next
    },
    dispose: () => reaction.dispose()
  }
})

// user-land mobx store
class Timer {
  secondsPassed = 0

  constructor() {
    makeAutoObservable(this)
  }

  increase() {
    this.secondsPassed += 1
  }

  reset() {
    this.secondsPassed = 0
  }
}

export const timer = new Timer();

setInterval(() => {
    timer.increase();
}, 1000);
```

In vue component:

```html
<script setup>
  import { timer } from './store.js'

  const doubled = computed(() => timer.secondPassed * 2) //even this is possible!
</script>
<template>
  <button @click="timer.reset()">
    Second passed: {{ timer.secondsPassed }}, doubled: {{ doubled }}
  </button>
</template>
```

# Motivation

It adds a lot of convinient, eliminates annoying wrappers (e.g. convert to a `ref`) and we can use third party reactive objects as if they are first-class citizen in vue.

This feature is primarily for __library authors__. Normal users do not need to worry about this.

# Detailed design

## API Summary
```ts
// pending deisgn
type InteropSource<T> = { track: ()=>T , dispose: ()=>void };
type InteropSourceFactory = <T>(fn: ()=>T, trigger: ()=> void) => InteropSource<T>;

function addReactivityInterop(factory: InteropSourceFactory): void;
```

## Implementation

Currently `addReactivityInterop` will perform __irreversible__ side effect: setting the current `InteropSourceFactory` (which should be a global variable).

Whenever a `ReactiveEffect` is constructed, the current `InteropSourceFactory` should be called with the original `fn` passed in `ReactiveEffect.constructor` and a `trigger` function which trigger the `ReactiveEffect` (to be scheduled/re-run). The return value is a `InteropSource`, whose `.track` should replace `.fn` of current `ReactiveEffect` and `.dispose()` should be called whenever current `ReactiveEffect` will be cleaned up. Essentially there are two steps:
* Hook `ReactiveEffect.fn`, so third party libraries get the control to perform their own dependency tracking logic. 
* Cleanup whenever appropriate, to avoid memory leak.

Note this feature implies the (external) dependencies should be transparently collected (by external) and the control should be __NOT__ inversed by external (it's vue taking control at first, calling the third party code to perform their dependency tracking). However, not all reactivity implementations give user the control (usually the control is inversed, e.g. `computed(getter)` whose `getter` is called by framework, not user directly, while mobx's `Reaction.track(fn)` is called by user, and runs immediately) or support auto (transparent/implicit) dependency tracking. 
> It's possible to implement _auto dependency tracking_ and/or _auto subscription_ based on a reactivity implementation which doesn't support these features, yet it's not the topic of this RFC.

### Workaround

In vue 3.2+ there is a work-around by patching `ReactiveEffect`

```ts

let currentInteropSourceFactory:InteropSourceFactory | null; // set by `addReactivityInterop`

Object.defineProperty(ReactiveEffect.prototype, 'fn', {
  get(this: ReactiveEffect & PatchedReactiveEffect) {
    return this._fn
  },
  set(this: ReactiveEffect & PatchedReactiveEffect, originalFn) {
    if(currentInteropSourceFactory) {
      if (this._fn) {
        this._interopSource?.dispose()
      } else {
        this._signal = ref(undefined)
      }
      this._interopSource = currentInteropSourceFactory(originalFn,()=>triggerRef(this._signal));
      this._fn = () => {
        this._signal.value; // track in vue
        return this._interopSource.track();
      } 
    } else {
      this._fn = originalFn;
    }
  }
})

const origianlStop = ReactiveEffect.prototype.stop

Object.defineProperty(ReactiveEffect.prototype, 'stop', {
  value: function (this: ReactiveEffect & PatchedReactiveEffect) {
    this._interopSource?.dispose();
    this._interopSource = null;
    origianlStop.call(this);
  }
});

interface PatchedReactiveEffect {
  _interopSource?: InteropSource;
  _fn?: Function;
  _signal?: Ref<any>;
}
```

## Composed `InteropSourceFactory`

If `addReactivityInterop` is called multiple times, the current `InteropSourceFactory` should be composed.

# Drawbacks

- A bit of magic (but acceptable in the context of vue)
- Add a bit of performance overhead in creation of `ReactiveEffect`

# Alternatives

N/A

# Adoption strategy

This is a new API and should not affect the existing code. It's also a low-level API only intended for advanced users and library authors.

# Unresolved questions

- Better API naming?

**Comment by 3Shain** [comment]
[Playground](https://sfc.vuejs.org/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8YnV0dG9uIEBjbGljaz1cInRpbWVyLnJlc2V0KClcIj5cbiAgICBTZWNvbmQgcGFzc2VkOiB7eyB0aW1lci5zZWNvbmRzUGFzc2VkIH19LCBkb3VibGVkOiB7eyBkb3VibGVkIH19XG4gIDwvYnV0dG9uPlxuPC90ZW1wbGF0ZT5cbjxzY3JpcHQgc2V0dXA+XG4gIC8vIGh0dHBzOi8vZ2l0aHViLmNvbS92dWVqcy9yZmNzL3B1bGwvNDIxXG4gIGltcG9ydCB7IHRpbWVyIH0gZnJvbSAnLi9zdG9yZS5qcyc7XG4gIGltcG9ydCB7IGNvbXB1dGVkIH0gZnJvbSAndnVlJztcbiAgXG4gIGNvbnN0IGRvdWJsZWQgPSBjb21wdXRlZCgoKSA9PiB0aW1lci5zZWNvbmRzUGFzc2VkICogMilcbiAgXG4gIC8vIHJlZnJlc2ggdGhlIGJyb3dzZXIgd2hlbiB5b3Ugc2VlICdDYW5ub3QgcmVkZWZpbmUgcHJvcGVydHk6IGZuJ1xuICAvLyBUT0RPOiBmaXggaXRcbjwvc2NyaXB0PiAiLCJpbXBvcnQtbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJ2dWVcIjogXCJodHRwczovL3NmYy52dWVqcy5vcmcvdnVlLnJ1bnRpbWUuZXNtLWJyb3dzZXIuanNcIixcbiAgICBcIm1vYnhcIjogXCJodHRwczovL2Nkbi5za3lwYWNrLmRldi9tb2J4XCJcbiAgfVxufSIsInNoaW0uanMiOiJpbXBvcnQgeyBSZWFjdGl2ZUVmZmVjdCwgdHJpZ2dlclJlZiwgcmVmIH0gZnJvbSAndnVlJztcblxubGV0IGN1cnJlbnRGYWN0b3J5O1xuXG5leHBvcnQgZnVuY3Rpb24gYWRkUmVhY3Rpdml0eUludGVyb3AoZmFjdG9yeSkge1xuICBjdXJyZW50RmFjdG9yeSA9IGZhY3Rvcnk7XG4gIC8vIEZJWE1FOiBDYW5ub3QgcmVkZWZpbmUgcHJvcGVydHk6IGZuXG4gIE9iamVjdC5kZWZpbmVQcm9wZXJ0eShSZWFjdGl2ZUVmZmVjdC5wcm90b3R5cGUsICdmbicsIHtcbiAgZ2V0KCkge1xuICAgIHJldHVybiB0aGlzLl9mblxuICB9LFxuICBzZXQob3JpZ2luYWxGbikge1xuICAgIGlmKGN1cnJlbnRGYWN0b3J5KSB7XG4gICAgICBpZiAodGhpcy5fZm4pIHtcbiAgICAgICAgdGhpcy5faW50ZXJvcFNvdXJjZT8uZGlzcG9zZSgpXG4gICAgICB9IGVsc2Uge1xuICAgICAgICB0aGlzLl9zaWduYWwgPSByZWYodW5kZWZpbmVkKVxuICAgICAgfVxuICAgICAgdGhpcy5faW50ZXJvcFNvdXJjZSA9IGN1cnJlbnRGYWN0b3J5KG9yaWdpbmFsRm4sKCk9PnRyaWdnZXJSZWYodGhpcy5fc2lnbmFsKSk7XG4gICAgICB0aGlzLl9mbiA9ICgpID0+IHRoaXMuX2ludGVyb3BTb3VyY2UudHJhY2sodGhpcy5fc2lnbmFsLnZhbHVlKTtcbiAgICB9IGVsc2Uge1xuICAgICAgdGhpcy5fZm4gPSBvcmlnaW5hbEZuO1xuICAgIH1cbiAgfVxufSlcblxuY29uc3Qgb3JpZ2lhbmxTdG9wID0gUmVhY3RpdmVFZmZlY3QucHJvdG90eXBlLnN0b3BcblxuT2JqZWN0LmRlZmluZVByb3BlcnR5KFJlYWN0aXZlRWZmZWN0LnByb3RvdHlwZSwgJ3N0b3AnLCB7XG4gIHZhbHVlOiBmdW5jdGlvbiAoKSB7XG4gICAgdGhpcy5faW50ZXJvcFNvdXJjZT8uZGlzcG9zZSgpO1xuICAgIHRoaXMuX2ludGVyb3BTb3VyY2UgPSBudWxsO1xuICAgIG9yaWdpYW5sU3RvcC5jYWxsKHRoaXMpO1xuICB9XG59KTtcbn0iLCJzdG9yZS5qcyI6Ii8vIHVzZSBtb2J4IGFzIGEgZGVtb3N0cmF0aW9uIG9mIGV4dGVybmFsIHJlYWN0aXZpdHkgc291cmNlXG5pbXBvcnQgeyBSZWFjdGlvbiwgbWFrZUF1dG9PYnNlcnZhYmxlIH0gZnJvbSAnbW9ieCdcbmltcG9ydCB7IGFkZFJlYWN0aXZpdHlJbnRlcm9wIH0gZnJvbSAnLi9zaGltLmpzJ1xuXG4vLyByZWdpc3RlciBNb2JYXG4vLyB0aGlzIHN0ZXAgc2hvdWxkIHVzdWFsbHkgYmUgcGVyZm9ybWVkIGJ5IF9saWJyYXJ5IGF1dGhvcnNfLlxubGV0IGlkID0gMFxuYWRkUmVhY3Rpdml0eUludGVyb3AoKGZuLCB0cmlnZ2VyKSA9PiB7XG4gIGNvbnN0IHJlYWN0aW9uID0gbmV3IFJlYWN0aW9uKGBleHRlcm5hbFNvdXJjZUAkeysraWR9YCwgdHJpZ2dlcilcbiAgcmV0dXJuIHtcbiAgICB0cmFjazogKHgpID0+IHtcbiAgICAgIGxldCBuZXh0XG4gICAgICByZWFjdGlvbi50cmFjaygoKSA9PiAobmV4dCA9IGZuKHgpKSlcbiAgICAgIHJldHVybiBuZXh0XG4gICAgfSxcbiAgICBkaXNwb3NlOiAoKSA9PiByZWFjdGlvbi5kaXNwb3NlKClcbiAgfVxufSlcblxuLy8gdXNlci1sYW5kIG1vYnggc3RvcmVcbmNsYXNzIFRpbWVyIHtcbiAgc2Vjb25kc1Bhc3NlZCA9IDBcblxuICBjb25zdHJ1Y3RvcigpIHtcbiAgICBtYWtlQXV0b09ic2VydmFibGUodGhpcylcbiAgfVxuXG4gIGluY3JlYXNlKCkge1xuICAgIHRoaXMuc2Vjb25kc1Bhc3NlZCArPSAxXG4gIH1cblxuICByZXNldCgpIHtcbiAgICB0aGlzLnNlY29uZHNQYXNzZWQgPSAwXG4gIH1cbn1cblxuZXhwb3J0IGNvbnN0IHRpbWVyID0gbmV3IFRpbWVyKCk7XG5cbnNldEludGVydmFsKCgpID0+IHtcbiAgICB0aW1lci5pbmNyZWFzZSgpO1xuICB9LCAxMDAwKTsifQ==)

A similar implementation (not in vue), for reference: solidjs/solid#739

**Comment by jods4** [comment]
Wouldn't that problem be solved more simply by making `track` and `trigger` (from `@vue/reactivity`) public APIs?

As a bonus side-effect it would enable creating custom/more efficient reactive objects/primitives; and it would allow other reactive systems such as MobX to integrate into Vue reactivity.

I have proposed this in the past and Evan was opposed to making `trigger` and `track` stable, public APIs; but this proposal is also exposing the trigger and track concepts (under those exact names), wrapped inside different objects, so it's really not any higher-level.

(Note: `trigger` and `track` could actually be used today; it's just that Evan said it's not officially a public API and stability is not guaranteed for future releases)

**Reply by 3Shain** [reply]
Please elaborate how the problem is solved by `track` and `trigger`. Let's implement an equivalent of the demo I posted above. I want to know how could you handle cleanup logic, as there exists implicit subscriptions (how do you know a reactive object is untracked?).

I would agree the `trigger` is the same, but `track` in this proposal is related to external dependency tracking behavior. It could be also named `patchedFn` which looks not lovely. So there could be multiple dependency tracking mechanism co-existing. While what you suggested is that let vue be the only one managing dependencies. This is hard (need a stable API, a sophisticated internal abstraction. I doubt if the abstraction is going to leak) and __intrusive__ (when to call vue's `track`? Of course it's up to the external. But how could a framework-agnostic implementation like mobx know how and when to call it?)

**Reply by jods4** [reply]
I'm not suggesting that Vue would be the only one managing dependencies.

I'm proposing that anyone could create objects other than ref/reactive that can _integrate_ with Vue tracking. Exposing `track` and `trigger` seems everything needed for that. MobX primitives would be one application of this.

So how would MobX make its primitives directly usable in Vue code?

Mobx basically works the same as Vue, internally it has [`reportObserved`](https://github.com/mobxjs/mobx/blob/main/packages/mobx/src/core/observable.ts#L134) and [`propagateChange`](https://github.com/mobxjs/mobx/blob/main/packages/mobx/src/core/observable.ts#L184).

When `reportObserved` is called, if there's no MobX context to observe the read (i.e. we're not running inside a MobX computed or reaction) then MobX can call `track`, which would register whatever MobX state is presently being read as a dependency for some potential Vue context (i.e. we're running inside a Vue effect).

When `propagateChange` is called, MobX can additionally call `trigger`, so that Vue effects that may depend on this dependency can also run. Scheduling would of course be separate and I don't quite see a good way to merge them.

I didn't completely think this through, do you see any reason it wouldn't work?

**Reply by 3Shain** [reply]
> Exposing track and trigger seems everything needed for that.

Just wanna say it's not true. Prove it with your code. Do not say "seems". I have given a working demo above.

> then MobX can call track ...
> When propagateChange is called, MobX can additionally call trigger ...

How could you make this possible given MobX is __framework-agnostic__ and it haven't provide such 'hooks' where you can call `track` and `trigger`?

> I didn't completely think this through, do you see any reason it wouldn't work?

I've clearly stated above: how would you manage subscription? When should external sources be unsubscribe to avoid memory leaking? Unless `track` receives an `onInvalidate` callback, but that's not how vue reactivity be designed.

**Reply by jods4** [reply]
> Just wanna say it's not true. 

I think it is, that being said I see from your next point an important design goal that you hadn't stated explicitly:

> How could you make this possible given MobX is framework-agnostic and it haven't provide such 'hooks' where you can call track and trigger?

OK so you want to integrate a lib _without_ an integration point inside the 3rd-party lib itself. 

In that case what I proposed isn't feasible: it does need a hook inside the 3rd party lib -- and I believe that would have been the most efficient design.
This surely is a hard ask for something like Knockout, but I recon MobX is actually designed to be integrated with other frameworks, so something that might happen?

Anyway, I'll start a new thread to discuss your idea.

**Reply by 3Shain** [reply]
> "think" "believe"

talk is cheap, show your code.

**Reply by jods4** [reply]
> talk is cheap, show your code.

The specs of the code are there: 
> When reportObserved is called, if there's no MobX context to observe the read (i.e. we're not running inside a MobX computed or reaction) then MobX can call track, which would register whatever MobX state is presently being read as a dependency for some potential Vue context (i.e. we're running inside a Vue effect).
>
> When propagateChange is called, MobX can additionally call trigger, so that Vue effects that may depend on this dependency can also run. Scheduling would of course be separate and I don't quite see a good way to merge them.

It's basically two small insertions of calls to `track` and `trigger`. I hope you can figure them out by yourself.

**Reply by 3Shain** [reply]
> I've clearly stated above: how would you manage subscription? When should external sources be unsubscribe to avoid memory leaking?

Answer the questions.

**Reply by jods4** [reply]
> Answer the questions.
>
> I've clearly stated above: how would you manage subscription? When should external sources be unsubscribe to avoid memory leaking?

Vue manages them. What I proposed was purely based on `track` + `trigger`.
Subscriptions are kept alive as long as the `ReactiveEffect` holding them is.
This is no different from how "internal" Vue dependencies are managed.

**Reply by 3Shain** [reply]
I hope you have really tried to implement it, but not just make assumption based on not-so-formal abstractions. 

The dependencies may change if there are branches inside `ReactiveEffect.fn`, how to deal with old dependencies which are not used? Always keeping then until `ReactiveEffect` disposed? If yes, I have no clue how it is efficient.

**Reply by 3Shain** [reply]
And how should vue notify mobx that some dependencies are not used? You must have a 'lifecycle' model. That's another abstraction which `track` doesn't provide.

**Reply by 3Shain** [reply]
Vue's `ReactiveEffect` lifecycle is managed by `EffectScope`, but mobx implements __on-demand subscription__ (that there is no explicit subscription, but a 'atom' is alive if and only if it's observed by a Reaction/Computation, and `Reaction` is the only abstraction that has a explicit `dispose` method), the latter is much more sophisticated.

**Reply by 3Shain** [reply]
> It's basically two small insertions of calls to track and trigger. I hope you can figure them out by yourself.

small insertions -> they work the same

what's your logic? I don't require you to prove by formal logic but just provide your code. I have tested it out. Seems you haven't.

**Reply by jods4** [reply]
> The dependencies may change if there are branches inside ReactiveEffect.fn, how to deal with old dependencies which are not used? Always keeping then until ReactiveEffect disposed? If yes, I have no clue how it is efficient.

No. It works like internal dependencies do, because it's based on the same primitives: `track` and `trigger`, which are the core contract of what a "reactive" piece of state needs to implement inside Vue.
When an effect runs, if a previous dependency doesn't call `track` again, it's dropped from this effect dependencies list.
This is how Vue works.

> And how should vue notify mobx that some dependencies are not used? You must have a 'lifecycle' model. That's another abstraction which track doesn't provide.

Dependencies tracked by Vue through `track` are not tracked by MobX (unless they are part of a separate `Reaction` of course). There is no MobX "lifecycle" required.

---

BTW, I noticed that you _can_ patch into MobX, although it's hacky.
When running an effect, Vue could push its own `trackingContext` and `trackingDerivation` into `globalState` (which is accessible as `_getGlobalState` is exported by MobX) -- and provide "alternative" contexts and derivations based on its own internal `track` and `trigger`.

This would be super efficient, but creating a package to achieve this depends on two things:
1. Vue should expose a few APIs: a hook when an effect runs/stops (alt. one could monkey-patch `ReactiveEffect.prototype.run`) and `track`/`trigger` (already exposed but not officially supported).
2. Custom `trackingContext` and `trackingDerivation` will depend a fair bit on MobX internals. Although if this was a package provided by MobX in the same way it supports React (e.g. `mobx-vue`), I guess the dependency on internal shapes is acceptable and it would be maintained along MobX itself.

**Reply by 3Shain** [reply]
> it's dropped from this effect dependencies list.

but how does a mobx's computation know it's alive? If it's not alive, it __will be never propagated by upstream changes__.

**Reply by 3Shain** [reply]
```js
import { makeObservable, observable, computed, autorun } from "mobx"

class OrderLine {
    price = 0
    amount = 1

    constructor(price) {
        makeObservable(this, {
            price: observable,
            amount: observable,
            total: computed
        })
        this.price = price
    }

    get total() {
        return this.price * this.amount
    }
}

const order = new OrderLine(0)
```

how is `order.total` be tracked without setting-up a subscription? If it's never subscribed, it will be never notified by the change from `price` & `amount`. Unlike vue's `computed` is always alive given it's always created under an `EffectScope`. 

Indirectly tracking `price` & `amount` (as they are the 'sources' and don't require subscription)? Then mobx's computation is just useless. (but it should be useful, because of on-demand subscription & memo)

**Reply by jods4** [reply]
It's a good point, that make the complete solution a little more involved but still doable.

The problem of course is not to start the computation, which works easily enough, but to be able to stop it as Vue doesn't have a public API for dropping effect dependencies.

I think the most straightforward way to do this is to let MobX do the bookkeeping.

Based on my most recent idea, i.e. create custom `trackingContext` and `trackingDerivation`:
If we base `vueContext` and `vueDerivation` on the existing Mobx code, it already does all the bookkeeping.
Those context and derivation only need to be modified to add extra `track` / `trigger` calls.

Concretely the code changes compared to the existing contexts and derivations would be:
- Call Vue `track` when calling `addObserver` in the derivation;
- The derivation `onBecomeStale_` should call `trigger`.

----
EDIT: a few considerations on this idea: 
1. There is a double book-keeping going on when a MobX object becomes a Vue dependency, as both MobX and Vue keep track of dependencies. To avoid this, Vue would have to make its dependency management public, I guess at least a hook when a dependency is removed from an effect (not sure it's worth it).
2. The naive implementation described here creates a context + derivation for every Vue effectScope, even when most of them wouldn't need it. This could be avoided by making them lazy: monkey-patch properties on `globalState` that only create them lazily when a MobX object was called, so no work is done when an effect runs 100% in Vue code.

**Reply by 3Shain** [reply]
That's different from your initial idea. 

Anyway __non-intrusiveness is my bottom line__. Perf impact should be experimented by benchmarks. There are trade-offs (and vue is essentially _an art of trade-offs_), you gain something then you lose something. This proposal is not perf oriented. If someone really need great perf, they shouldn't consider interoperability too much.

**Reply by jods4** [reply]
> That's different from your initial idea.

Not that much.
The minute details of where the code goes are different, but code itself is not that important.

My fundamental question was: aren't `trigger` / `track` enough? 

To be transparent, I have my own motivations for asking this question:
I've mentioned it before but I think you can do many interesting things when given access to `track` and `trigger`, outside the scope of interop with other libs. There's currently another discussion that basically amounts to exposing the shape of `track` and `trigger`: #364 

Minimizing the amount of public API when possible is definitely a good idea.

**Reply by 3Shain** [reply]
My conclusion is 'not enough' but I have no responsibility to prove it :). Please tell me if you find one which is not intrusive.

P.S. I will not reply a word until there is a real implementation (a playground or a repo)

**Comment by jods4** [comment]
There are two things that I don't find elegant in your proposal:

1. My understanding is that it only works with a single 3rd party lib. 
For example if I'm migrating to Vue a legacy app that was written in Knockout with MobX added, then I couldn't hook both KO and MobX into this system.

2. It is taxing perf and memory-wise.
Effects are at the heart of Vue, in a large app there are _lots_ of them: one for every component, one for every computed (and man some libs, e.g. validation, create _tons_ of computed).
There was some effort into optimizing them.
Once a 3rd party is integrated, _each_ effect will run inside a 3rd party "effect" (`Reaction` in MobX).

Sadly, I don't see how to improve those points without having the 3rd party lib cooperate :(

**Reply by 3Shain** [reply]
1.
> Composed InteropSourceFactory
> If addReactivityInterop is called multiple times, the current InteropSourceFactory should be composed.

2. 
True. Not able to solve until __there is a working alternative__
